File: src/text.cpp
/*
    Plugin for computers in vanilla minecraft!
    Copyright (C) 2022  JNNGL

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "j_env.h"
#include "utils.h"
#include "text.h"

using namespace text;

// java/text/CharacterIterator
// Name: CharacterIterator
// Base Class(es): [NativeObjectInstance]

NativeObject clsCharacterIterator;
jclass InitCharacterIterator() {
   return tcInitNativeObject(clsCharacterIterator, "java/text/CharacterIterator", { 
          { "clone", "()Ljava/lang/Object;", false }, 
          { "current", "()C", false }, 
          { "first", "()C", false }, 
          { "getBeginIndex", "()I", false }, 
          { "getEndIndex", "()I", false }, 
          { "getIndex", "()I", false }, 
          { "last", "()C", false }, 
          { "next", "()C", false }, 
          { "previous", "()C", false }, 
          { "setIndex", "(I)C", false }
       }, { 
       });
}

CharacterIterator::CharacterIterator(jobject instance)
 : instance(instance), NativeObjectInstance(instance) {
    InitCharacterIterator();
}

CharacterIterator::~CharacterIterator() {}

lang::Object CharacterIterator::Clone() {
   return (lang::Object)(tc::jEnv->CallObjectMethod(METHOD(clsCharacterIterator, 0)));
}

signed char CharacterIterator::Current() {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 1));
}

signed char CharacterIterator::First() {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 2));
}

int CharacterIterator::GetBeginIndex() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsCharacterIterator, 3));
}

int CharacterIterator::GetEndIndex() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsCharacterIterator, 4));
}

int CharacterIterator::GetIndex() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsCharacterIterator, 5));
}

signed char CharacterIterator::Last() {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 6));
}

signed char CharacterIterator::Next() {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 7));
}

signed char CharacterIterator::Previous() {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 8));
}

signed char CharacterIterator::SetIndex(int position) {
   return (signed char)tc::jEnv->CallByteMethod(METHOD(clsCharacterIterator, 9), (jint)position);
}


// java/text/AttributedCharacterIterator
// Name: AttributedCharacterIterator
// Base Class(es): [CharacterIterator]

NativeObject clsAttributedCharacterIterator;
jclass InitAttributedCharacterIterator() {
   return tcInitNativeObject(clsAttributedCharacterIterator, "java/text/AttributedCharacterIterator", { 
          { "getAllAttributeKeys", "()Ljava/util/Set;", false }, 
          { "getAttribute", "(Ljava/text/AttributedCharacterIterator$Attribute;)Ljava/lang/Object;", false }, 
          { "getAttributes", "()Ljava/util/Map;", false }, 
          { "getRunLimit", "()I", false }, 
          { "getRunLimit", "(Ljava/text/AttributedCharacterIterator$Attribute;)I", false }, 
          { "getRunLimit", "(Ljava/util/Set;)I", false }, 
          { "getRunStart", "()I", false }, 
          { "getRunStart", "(Ljava/text/AttributedCharacterIterator$Attribute;)I", false }, 
          { "getRunStart", "(Ljava/util/Set;)I", false }
       }, { 
       });
}

AttributedCharacterIterator::AttributedCharacterIterator(jobject instance)
 : instance(instance), CharacterIterator(instance) {
    InitAttributedCharacterIterator();
}

AttributedCharacterIterator::~AttributedCharacterIterator() {}

util::Set AttributedCharacterIterator::GetAllAttributeKeys() {
   return (util::Set)(tc::jEnv->CallObjectMethod(METHOD(clsAttributedCharacterIterator, 0)));
}

lang::Object AttributedCharacterIterator::GetAttribute(AttributedCharacterIterator::Attribute attribute) {
   return (lang::Object)(tc::jEnv->CallObjectMethod(METHOD(clsAttributedCharacterIterator, 1), attribute.instance));
}

util::Map AttributedCharacterIterator::GetAttributes() {
   return (util::Map)(tc::jEnv->CallObjectMethod(METHOD(clsAttributedCharacterIterator, 2)));
}

int AttributedCharacterIterator::GetRunLimit() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 3));
}

int AttributedCharacterIterator::GetRunLimit(AttributedCharacterIterator::Attribute attribute) {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 4), attribute.instance);
}

int AttributedCharacterIterator::GetRunLimit(util::Set attributes) {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 5), attributes.instance);
}

int AttributedCharacterIterator::GetRunStart() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 6));
}

int AttributedCharacterIterator::GetRunStart(AttributedCharacterIterator::Attribute attribute) {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 7), attribute.instance);
}

int AttributedCharacterIterator::GetRunStart(util::Set attributes) {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator, 8), attributes.instance);
}


// java/text/AttributedCharacterIterator/Attribute
// Name: AttributedCharacterIterator::Attribute
// Base Class(es): [lang::Object]

NativeObject clsAttributedCharacterIterator_Attribute;
jclass InitAttributedCharacterIterator_Attribute() {
   return tcInitNativeObject(clsAttributedCharacterIterator_Attribute, "java/text/AttributedCharacterIterator/Attribute", { 
          { "equals", "(Ljava/lang/Object;)Z", false }, 
          { "hashCode", "()I", false }, 
          { "toString", "()Ljava/lang/String;", false }
       }, { 
          { "INPUT_METHOD_SEGMENT", "Ljava/text/AttributedCharacterIterator$Attribute;", true }, 
          { "LANGUAGE", "Ljava/text/AttributedCharacterIterator$Attribute;", true }, 
          { "READING", "Ljava/text/AttributedCharacterIterator$Attribute;", true }
       });
}

AttributedCharacterIterator::Attribute::Attribute(jobject instance)
 : instance(instance), lang::Object(instance) {
    InitAttributedCharacterIterator_Attribute();
}

AttributedCharacterIterator::Attribute::~Attribute() {}

bool AttributedCharacterIterator::Attribute::Equals(lang::Object obj) {
   return (bool)tc::jEnv->CallBooleanMethod(METHOD(clsAttributedCharacterIterator_Attribute, 0), obj.instance);
}

int AttributedCharacterIterator::Attribute::HashCode() {
   return (int)tc::jEnv->CallIntMethod(METHOD(clsAttributedCharacterIterator_Attribute, 1));
}

char* AttributedCharacterIterator::Attribute::ToString() {
   return FromJString((jstring)tc::jEnv->CallObjectMethod(METHOD(clsAttributedCharacterIterator_Attribute, 2)));
}

AttributedCharacterIterator::Attribute AttributedCharacterIterator::Attribute::INPUT_METHOD_SEGMENT() {
   InitAttributedCharacterIterator_Attribute();
   return (AttributedCharacterIterator::Attribute)(tc::jEnv->GetStaticObjectField(clsAttributedCharacterIterator_Attribute.clazz, clsAttributedCharacterIterator_Attribute.fields[0]));
}

AttributedCharacterIterator::Attribute AttributedCharacterIterator::Attribute::LANGUAGE() {
   InitAttributedCharacterIterator_Attribute();
   return (AttributedCharacterIterator::Attribute)(tc::jEnv->GetStaticObjectField(clsAttributedCharacterIterator_Attribute.clazz, clsAttributedCharacterIterator_Attribute.fields[1]));
}

AttributedCharacterIterator::Attribute AttributedCharacterIterator::Attribute::READING() {
   InitAttributedCharacterIterator_Attribute();
   return (AttributedCharacterIterator::Attribute)(tc::jEnv->GetStaticObjectField(clsAttributedCharacterIterator_Attribute.clazz, clsAttributedCharacterIterator_Attribute.fields[2]));
}

File: src/main/java/com/jnngl/totalcomputers/system/ui/Text.java
/*
    Plugin for computers in vanilla minecraft!
    Copyright (C) 2022  JNNGL

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.jnngl.totalcomputers.system.ui;

import com.jnngl.totalcomputers.system.Utils;

import java.awt.*;

/**
 * Text UI Element
 */
public class Text {

    private int x, y, width, height;
    private Font font;
    private Color color;
    private String text;
    protected int boundsWidth, boundsHeight;

    /**
     * Constructor
     * @param x X coordinate
     * @param y Y coordinate
     * @param width Maximum width
     * @param height Maximum height
     * @param font Font
     * @param color Color
     * @param text Text
     */
    public Text(int x, int y, int width, int height, Font font, Color color, String text) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.font = font;
        this.color = color;
        updateText(text);
    }

    /**
     * Constructor
     * @param x X coordinate
     * @param y Y coordinate
     * @param font Font
     * @param color Color
     * @param text Text
     */
    public Text(int x, int y, Font font, Color color, String text) {
        this(x, y, Integer.MAX_VALUE, Integer.MAX_VALUE, font, color, text);
    }

    /**
     * Renders the text
     * @param g Graphics2D instance
     */
    public void render(Graphics2D g) {
        g.setColor(color);
        g.setFont(font);
        FontMetrics metrics = g.getFontMetrics();
        int height = metrics.getHeight();
        int y = this.y-height;
        for (String line : text.split("\n"))
            g.drawString(line, x, y += height);
    }

    /**
     * Updates the text layout
     * @param text Source text
     */
    private void updateText(String text) {
        this.text = "";
        int line = 1;
        FontMetrics m = Utils.getFontMetrics(font);
        int fHeight = m.getHeight();
        if(m.stringWidth(text) < width) {
            this.text = text;
        } else {
            String[] words = text.split(" ");
            StringBuilder formatted = new StringBuilder();
            StringBuilder currentLine = new StringBuilder(words[0]);
            for(int i = 1; i < words.length; i++) {
                if(m.stringWidth(currentLine+words[i]) < width) {
                    currentLine.append(" ").append(words[i]);
                } else {
                    line++;
                    if(line*fHeight > height) {
                        if(formatted.length() > 3) {
                            this.text = formatted.substring(0, formatted.length()-3);
                            this.text += "...";
                            return;
                        }
                    }
                    currentLine.append("\n");
                    formatted.append(currentLine);
                    currentLine = new StringBuilder(words[i]);
                }
            }
            if(currentLine.toString().trim().length() > 0) {
                formatted.append(currentLine);
            }
            this.text = formatted.toString();
        }
        FontMetrics metrics = Utils.getFontMetrics(font);
        boundsHeight = 0;
        boundsWidth = 0;
        for (String _line : text.split("\n")) {
            int strWidth = metrics.stringWidth(_line);
            boundsWidth = Math.max(boundsWidth, strWidth);
            boundsHeight += metrics.getHeight();
        }
    }

    /**
     * Getter
     * @return this.x
     */
    public int getX() {
        return x;
    }

    /**
     * Setter
     * @param x New X
     */
    public void setX(int x) {
        this.x = x;
    }

    /**
     * Getter
     * @return this.y
     */
    public int getY() {
        return y;
    }

    /**
     * Setter
     * @param y New Y
     */
    public void setY(int y) {
        this.y = y;
    }

    /**
     * Getter
     * @return this.width
     */
    public int getWidth() {
        return width;
    }

    /**
     * Setter
     * @param width New width
     */
    public void setWidth(int width) {
        this.width = width;
        updateText(text);
    }

    /**
     * Getter
     * @return this.height
     */
    public int getHeight() {
        return height;
    }

    /**
     * Setter
     * @param height New height
     */
    public void setHeight(int height) {
        this.height = height;
        updateText(text);
    }

    /**
     * Getter
     * @return this.font
     */
    public Font getFont() {
        return font;
    }

    /**
     * Setter
     * @param font New font
     */
    public void setFont(Font font) {
        this.font = font;
        updateText(text);
    }

    /**
     * Getter
     * @return this.color
     */
    public Color getColor() {
        return color;
    }

    /**
     * Setter
     * @param color New color
     */
    public void setColor(Color color) {
        this.color = color;
    }

    /**
     * Getter
     * @return this.text
     */
    public String getText() {
        return text;
    }

    /**
     * Setter
     * @param text New text
     */
    public void setText(String text) {
        this.text = text;
        updateText(text);
    }
}
